io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/Schedule.java

```diff
@@ -0,0 +1,492 @@
+			case DELAY_DISCHARGE:
+				if (this.isDelayDischargeScheduled()) {
+					stateMachine = StateMachine.DELAYED;
+				}
+				break;
+			}
+
+			return stateMachine;
+		}
+	}
+
+	public Schedule(ControlMode controlMode, int essUsableEnergy, int currentAvailableEnergy, //
+			int dischargeEnergy, int chargeEnergy, Float[] prices, Integer[] consumptionPrediciton, //
+			Integer[] productionPrediction, int maxAllowedChargeEnergyFromGrid) {
+
+		this.controlMode = controlMode;
+		this.essUsableEnergy = essUsableEnergy;
+		this.currentAvailableEnergy = currentAvailableEnergy;
+		this.maxDischargeEnergyPerPeriod = dischargeEnergy;
+		this.maxChargeEnergyPerPeriod = chargeEnergy;
+		this.maxAllowedChargeEnergyFromGrid = maxAllowedChargeEnergyFromGrid;
+
+		// Filtering non null values.
+		var priceValues = Arrays.stream(prices) //
+				.filter(Objects::nonNull) //
+				.toArray(Float[]::new);
+
+		for (var index = 0; index < priceValues.length; index++) {
+			var consumption = consumptionPrediciton[index];
+			var production = productionPrediction[index];
+			var price = priceValues[index];
+
+			if (production == null) {
+				production = 0;
+			}
+
+			if (consumption == null) {
+				consumption = 0;
+			}
+
+			// Creating the period.
+			// Initially the 'chargeDischarge','grid' and 'initialEnergy' will be 'null'.
+			var period = new Period(null, null, null, price, production, consumption, this.maxDischargeEnergyPerPeriod);
+
+			// Adding to the list.
+			this.periods.add(period);
+		}
+
+		// Creates the initial schedule based on balancing.
+		this.simulateSchedule();
+	}
+
+	/**
+	 * Creates the schedule with periods.
+	 */
+	public void createSchedule() {
+		for (int index = 0; index < this.periods.size(); index++) {
+			var period = this.periods.get(index);
+			var requiredEnergy = period.requiredEnergy;
+			var essInitialEnergy = period.essInitialEnergy;
+
+			if (requiredEnergy >= essInitialEnergy) {
+				// recalculate the required energy needed.
+				var requiredEnergyFromGrid = requiredEnergy - essInitialEnergy;
+				if (requiredEnergyFromGrid > 0) {
+					this.calculateSchedule(requiredEnergyFromGrid, index);
+				}
+			}
+			// Simulates and updates the schedule.
+			this.simulateSchedule();
+		}
+	}
+
+	/**
+	 * Simulates and updates the schedule.
+	 * 
+	 * <p>
+	 * Update in charge or discharge energy in a certain period changes the
+	 * subsequent periods.
+	 * 
+	 */
+	private void simulateSchedule() {
+		var essInitialEnergy = this.currentAvailableEnergy;
+
+		for (int index = 0; index < this.periods.size(); index++) {
+
+			// Current period.
+			var period = this.periods.get(index);
+
+			// calculate initial energy for current period.
+			final int essInitialEnergyForCurrentPeriod = TypeUtils.max(0, essInitialEnergy);
+
+			// Update the value.
+			period.essInitialEnergy = essInitialEnergyForCurrentPeriod;
+
+			// updates 'chargeDischarge energy' and 'Grid energy' values.
+			this.updateEssAndGridEnergyInPeriod(index);
```

Comment:
What's wrong with Michaels suggestion? You don't need 'index' in `updateEssAndGridEnergyInPeriod()`, so passing period seems to be better. 

(It's always a good idea to have strong typing (i.e. Period) instead of generic types like an int)